### PR TITLE
[KOGITO-1958] Implement decision tracing addon for Quarkus

### PR DESCRIPTION
**Pull request by kostola**

### CODE REVIEW

File: kogito-codegen/src/main/resources/class-templates/DMNApplicationClassDeclTemplate.java [hunk 1]
```diff
@@ -2,13 +2,21 @@
...
+        }
+    }
+
+    private boolean requiresExecutionIdSupplier(org.kie.kogito.Application app) {
+        return app.config().addons().availableAddons().contains("tracing-decision");
     }
 
     public org.kie.kogito.decision.DecisionModel getDecisionModel(java.lang.String namespace, java.lang.String name) {
-        return new org.kie.kogito.dmn.DmnDecisionModel(dmnRuntime, namespace, name);
+        return new org.kie.kogito.dmn.DmnDecisionModel(dmnRuntime, namespace, name, execIdSupplier);
```
Comment (tarilabs) [inline kogito-codegen/src/main/resources/class-templates/DMNApplicationClassDeclTemplate.java:19]: not 100% what is the decision from Trusty side of things, but shouldn't this type of configuration be a compile-time decision?

Comment (kostola) [reply kogito-codegen/src/main/resources/class-templates/DMNApplicationClassDeclTemplate.java:19]: Definitely. It's implemented this way at the moment because it was the quickest to have the addon working to do local tests. I will change it for sure before the final review.

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/DecisionTracingCollector.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            return null;
+        }
+    }
+
+    public static class CloudEvent<T> {
+        private final String specversion;
+        private final String id;
+        private final String type;
+        private final URI source;
+        private final T data;
```
Comment (r00ta) [inline addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/DecisionTracingCollector.java:154]: Why this? io.cloudevents already provide a builder for the cloud event `import io.cloudevents.v1.CloudEventBuilder;`

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        }
+
+        public List<DecisionResult> getDecisionResults() {
+            return decisionResults;
+        }
+
+        public List<Message> getMessages() {
+            return messages;
+        }
+    }
```
Comment (r00ta) [inline addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java:85]: Move to another file

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        private final List<Message> messages;
+        private final boolean errors;
+
+        public DecisionResult(String decisionId, String decisionName, DMNDecisionResult.DecisionEvaluationStatus evaluationStatus, Object result, List<Message> messages, boolean errors) {
+            this.decisionId = decisionId;
+            this.decisionName = decisionName;
+            this.evaluationStatus = evaluationStatus;
+            this.result = result;
+            this.messages = messages;
+            this.errors = errors;
```
Comment (r00ta) [inline addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java:102]: Move to another file (also for all the other files/classes)

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import org.kie.dmn.api.core.DMNContext;
+import org.kie.dmn.api.core.DMNDecisionResult;
+import org.kie.dmn.api.core.DMNMessage;
+import org.kie.dmn.api.core.DMNMessageType;
+import org.kie.dmn.api.core.DMNResult;
+
+public abstract class EvaluateEvent implements Serializable {
```
Comment (danielezonca) [inline addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java:15]: This class will be probably changed a lot soon but in general I would prefer to have few/none inner classes, especially when it is not a sub-component of the outer class but a proper concept like `DecisionResult`, `Result` etc.
My proposal is to promote all of them as proper classes, wdyt?

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/DecisionTracingCollector.java (hunk 2 above)
Comment (kostola) [reply addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/DecisionTracingCollector.java]: When you wrote this comment the official CloudEvents SDK (that contains the package you mentioned) was not imported and this class was used as POJO for the event.
Now we decided to use it because of some useful extensions and this was removed.

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java (hunk 3 above)
Comment (kostola) [reply addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java]: I will definitely refactor this part once we have a clear agreement on the first version of the data we want to export.

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java (hunk 4 above)
Comment (kostola) [reply addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java]: I will definitely refactor this part once we have a clear agreement on the first version of the data we want to export.

File: addons/tracing/tracing-decision-common/pom.xml [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  <dependencies>
+    <dependency>
+      <groupId>org.kie.kogito</groupId>
+      <artifactId>kogito-dmn</artifactId>
+    </dependency>
+
+    <dependency>
+      <groupId>io.cloudevents</groupId>
+      <artifactId>cloudevents-api</artifactId>
+      <version>1.3.0</version>
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-common/pom.xml]: Done :+1:

File: addons/tracing/tracing-decision-common/pom.xml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-engine</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-nop</artifactId>
+      <version>${version.org.slf4j}</version>
+      <scope>test</scope>
+    </dependency>
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-common/pom.xml]: I realized this wasn't even needed

File: addons/tracing/tracing-decision-common/src/test/java/org/kie/kogito/tracing/decision/testimpl/TestAfterEvaluateAllEvent.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
+package org.kie.kogito.tracing.decision.testimpl;
+
+import org.kie.dmn.api.core.DMNResult;
+import org.kie.dmn.api.core.event.AfterEvaluateAllEvent;
+
+public class TestAfterEvaluateAllEvent implements AfterEvaluateAllEvent {
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-common/src/test/java/org/kie/kogito/tracing/decision/testimpl/TestAfterEvaluateAllEvent.java]: Done :+1:

File: addons/tracing/tracing-decision-quarkus-addon/pom.xml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-engine</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-nop</artifactId>
+      <version>${version.org.slf4j}</version>
+      <scope>test</scope>
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-quarkus-addon/pom.xml]: This wasn't needed here as well

File: addons/tracing/tracing-decision-quarkus-addon/src/test/java/org/kie/kogito/tracing/decision/DecisionTracingTest.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+        List<Pair<String, Object>> eventBusCalls = eventBus.getCalls();
+        assertEquals(2, eventBusCalls.size());
+        assertEquals("kogito-tracing-decision_BeforeEvaluateAllEvent", eventBusCalls.get(0).getLeft());
+        assertTrue(eventBusCalls.get(0).getRight() instanceof BeforeEvaluateAllEvent);
+        assertEquals("kogito-tracing-decision_AfterEvaluateAllEvent", eventBusCalls.get(1).getLeft());
+        assertTrue(eventBusCalls.get(1).getRight() instanceof AfterEvaluateAllEvent);
+
+        BeforeEvaluateAllEvent beforeEvent = (BeforeEvaluateAllEvent) eventBusCalls.get(0).getRight();
+        AfterEvaluateAllEvent afterEvent = (AfterEvaluateAllEvent) eventBusCalls.get(1).getRight();
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-quarkus-addon/src/test/java/org/kie/kogito/tracing/decision/DecisionTracingTest.java]: Done :+1:

File: addons/tracing/tracing-decision-common/src/test/java/org/kie/kogito/tracing/decision/testimpl/TestAfterEvaluateAllEvent.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
+package org.kie.kogito.tracing.decision.testimpl;
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-common/src/test/java/org/kie/kogito/tracing/decision/testimpl/TestAfterEvaluateAllEvent.java]: Done :+1:

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/aggregator/Aggregator.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+import java.util.List;
+
+import io.cloudevents.v1.CloudEventImpl;
+import org.kie.kogito.tracing.decision.event.EvaluateEvent;
+
+@FunctionalInterface
+public interface Aggregator<T> {
+
+    CloudEventImpl<T> aggregate(String evaluationId, List<EvaluateEvent> events);
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/aggregator/Aggregator.java:27]: The `v1` contained in the package is the version of the [CloudEvents specification v1.0](https://github.com/cloudevents/spec/blob/v1.0/spec.md), so it makes sense to me to have it there.

I used `CloudEventImpl` because it's the standard implementation. The `CloudEvent` interface seems too generic for me: it requires two type parameters, one for the attributes of the CloudEvent itself and one for the payload.
`CloudEventImpl`, instead, implements `CloudEvent` fixing the first type parameter to a standard implementation that matches the specification, and allows to specify only the payload type (the one we care about).
Idk why it's implemented this way but I assume for compatibility between different specs. I don't like the idea of using `CloudEvent` and specifying the attribute type everywhere to be honest.

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/AbstractDecisionTracingCollector.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        if (event instanceof AfterEvaluateAllEvent) {
+            String payload = aggregate(evaluationId, cacheMap.get(evaluationId));
+            handlePayload(payload);
+            LOG.debug("Generated aggregated event for evaluation {} (length {})", evaluationId, payload.length());
+            cacheMap.remove(evaluationId);
+            LOG.trace("Removed evaluation {} from cache (current size: {})", evaluationId, cacheMap.size());
+        }
+    }
+
+    protected abstract void handlePayload(String payload);
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/AbstractDecisionTracingCollector.java]: Done :+1:

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/AbstractDecisionTracingListener.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    public void afterEvaluateAll(org.kie.dmn.api.core.event.AfterEvaluateAllEvent event) {
+        handleEvaluateEvent(new AfterEvaluateAllEvent(
+                extractExecutionId(event.getResult().getContext()),
+                event.getModelName(),
+                event.getModelNamespace(),
+                event.getResult()
+        ));
+    }
+
+    protected abstract void handleEvaluateEvent(EvaluateEvent event);
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/AbstractDecisionTracingListener.java]: Done :+1:

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java (hunk 5 above)
Comment (kostola) [reply addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java:25]: Promoted all inner classes to top level Java classes :+1:

File: addons/tracing/tracing-decision-quarkus-addon/pom.xml [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+  <properties>
+    <maven.compiler.release>11</maven.compiler.release>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.kie.kogito</groupId>
+      <artifactId>kogito-tracing-decision-common</artifactId>
+      <version>${project.version}</version>
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-quarkus-addon/pom.xml]: I used `kogito-bom` like in `process-management` addon

File: kogito-codegen/src/main/resources/class-templates/DMNApplicationClassDeclTemplate.java (hunk 1 above)
Comment (kostola) [reply kogito-codegen/src/main/resources/class-templates/DMNApplicationClassDeclTemplate.java:12]: Now I'm checking if the addon is enabled at compile time and, if true, the instantiation of `execIdSupplier` is code-generated.

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/DecisionTracingListener.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.function.Consumer;
+
+import org.kie.dmn.api.core.DMNContext;
+import org.kie.dmn.api.core.event.DMNRuntimeEventListener;
+import org.kie.kogito.decision.DecisionExecutionIdUtils;
+import org.kie.kogito.tracing.decision.event.AfterEvaluateAllEvent;
+import org.kie.kogito.tracing.decision.event.BeforeEvaluateAllEvent;
+import org.kie.kogito.tracing.decision.event.EvaluateEvent;
+
+public class DecisionTracingListener implements DMNRuntimeEventListener {
```
Comment (r00ta) [inline addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/DecisionTracingListener.java:28]: Sorry for asking again, but is there a specific reason to export the information of the event using a listener? What are the additional information that we might get from it (that we don't have in the `DMNResult`?)

Comment (kostola) [reply addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/DecisionTracingListener.java:28]: We're using a listener because we already planned to increase the amount of exported information with additional data that can be obtained only via a listener (e.g. inputs/outputs of BKM model invocations)
Thus, even if in this version we're basically just exporting the final `DMNResult`, the implementation via listener already enables the next improvements.

File: kogito-quarkus-extension/deployment/src/main/java/org/kie/kogito/quarkus/deployment/KogitoAssetsProcessor.java [hunk 17]
```diff
@@ -75,7 +74,7 @@
     private final transient String appPackageName = "org.kie.kogito.app";
     private final transient String persistenceFactoryClass = "org.kie.kogito.persistence.KogitoProcessInstancesFactory";
     private final transient String metricsClass = "org.kie.kogito.monitoring.rest.MetricsResource";
-
+    private final transient String tracingClass = "org.kie.kogito.tracing.decision.QuarkusDecisionTracingListener";
```
Comment (danielezonca) [inline kogito-quarkus-extension/deployment/src/main/java/org/kie/kogito/quarkus/deployment/KogitoAssetsProcessor.java:77]: Please avoid `Quarkus` keyword here. This is the "marker" class we expect to have to recognize tracing is enable. Please rename it as `KogitoDecisionTracingListener` (of course rename the original class too, not only this variable :) )

Comment (kostola) [reply kogito-quarkus-extension/deployment/src/main/java/org/kie/kogito/quarkus/deployment/KogitoAssetsProcessor.java]: Done :+1:

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+
+package org.kie.kogito.tracing.decision;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import io.vertx.core.eventbus.EventBus;
+
+@ApplicationScoped
+public final class KogitoDecisionTracingListener extends DecisionTracingListener {
```
Comment (danielezonca) [inline addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java:25]: Can you please add a comment like [this one](https://github.com/kiegroup/kogito-runtimes/blob/master/addons/persistence/infinispan-persistence-addon/src/main/java/org/kie/kogito/persistence/KogitoProcessInstancesFactory.java#L27-L30)

Comment (kostola) [reply addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java:28]: Done :+1:

**Comment by danielezonca** [comment]
@MarianMacik @radtriste 
Can you please review this PR (feel free to assign a different QE if needed)?

@evacchi 
Do you want to review it too?

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java (hunk 18 above)
Comment (tarilabs) [reply addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java:28]: Can you put a link to the Doc about why "must always have" the exact FQN, please?

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+
+package org.kie.kogito.tracing.decision;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import io.vertx.core.eventbus.EventBus;
+
+/**
+ * This class must always have exact FQCN as <code>org.kie.kogito.tracing.decision.KogitoDecisionTracingListener</code>
```
Comment (tarilabs) [inline addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java:25]: Can you put a link to the Doc about why "must always have" the exact FQN, please?

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+        DMNContext clone = context.clone();
+        this.executionId = executionId;
+        this.modelName = modelName;
+        this.modelNamespace = modelNamespace;
+        this.context = clone.getAll();
+        this.contextMetadata = clone.getMetadata().asMap();
+        this.result = null;
+    }
+
+    public EvaluateEvent(String executionId, String modelName, String modelNamespace, DMNResult result) {
```
Comment (r00ta) [inline addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java:43]: Just a consideration from my side: this `EvaluateEvent` is collecting exacly the `DMNResult`. As we have already discussed in the past, if we need to export just the `DMNResult` I think we should do that in another way. 
But since we are using a listener 'cause potentially we might extract more information (that is not contained in the `DMNResult`), I'm a bit worried to expose here exactly the `DMNResult` structure. This is going to be deprecated very soon by design, just saying that nobody should base his implementation on the `DMNResult` exposed here.. wdyt?

File: addons/tracing/tracing-decision-common/src/test/resources/Traffic Violation.dmn [hunk 21]
```diff
@@ -0,0 +1,230 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (r00ta) [inline addons/tracing/tracing-decision-common/src/test/resources/Traffic Violation.dmn:1]: I see that this dmn is already present in another module of `kogito` (https://github.com/kiegroup/kogito-runtimes/blob/faf3780d9b9a2e004154f4810fa702cfefaa5e55/drools/kogito-dmn/src/test/resources/org/kie/kogito/dmn/TrafficViolation.dmn), is it ok to duplicate it in this module?

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingCollector.java [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+        eventSubject = PublishSubject.create();
+        collector = new DecisionTracingCollector(eventSubject::onNext);
+    }
+
+    @Outgoing("kogito-tracing-decision")
+    public Publisher<String> getEventPublisher() {
+        return eventSubject.toFlowable(BackpressureStrategy.BUFFER);
+    }
+
+    @ConsumeEvent("kogito-tracing-decision_BeforeEvaluateAllEvent")
```
Comment (r00ta) [inline addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingCollector.java:45]: I'm not an expert of vertx so I might ask a dumb question: is it possible that two events are handled at the same time by this method? Is there just one event loop that will manage it? i.e. we can't have a concurrency problem here right?

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/ApplicationGenerator.java [hunk 23]
```diff
@@ -204,6 +204,7 @@ public ApplicationGenerator withMonitoring(boolean monitoring) {
        return this;
    }
 
+
```
Comment (r00ta) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/ApplicationGenerator.java:207]: remove newline

File: kogito-maven-plugin/src/main/java/org/kie/kogito/maven/plugin/GenerateModelMojo.java [hunk 24]
```diff
@@ -206,7 +207,7 @@ private ApplicationGenerator createApplicationGenerator(
         }
 
         if (generateDecisions) {
-            appGen.withGenerator(DecisionCodegen.ofPath(kieSourcesDirectory.toPath()))
+            appGen.withGenerator(DecisionCodegen.ofPath(kieSourcesDirectory.toPath()).withTracing(useTracing))
```
Comment (r00ta) [inline kogito-maven-plugin/src/main/java/org/kie/kogito/maven/plugin/GenerateModelMojo.java:210]: Just a small consideration from my side: since we injecting some booleans as configurations, I think it's better to have a configuration object instead of many booleans, it would be much easier to maintain imo. If this is fine for the others we might do this refactoring in another PR

File: pom.xml [hunk 25]
```diff
@@ -752,6 +753,13 @@
         <version>${version.io.fabric8.kubernetes-client}</version>
         <scope>test</scope>
       </dependency>
+
+      <!-- CloudEvents -->
+      <dependency>
```
Comment (r00ta) [inline pom.xml:758]: Is this used somewhere?

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java (hunk 19 above)
Comment (MarianMacik) [reply addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java:25]: +1

Comment (danielezonca) [reply addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java:25]: I agree that this should be documented but it is following the same "convention" of other classes like [this one](https://github.com/kiegroup/kogito-runtimes/blob/master/addons/persistence/infinispan-persistence-addon/src/main/java/org/kie/kogito/persistence/KogitoProcessInstancesFactory.java#L27-L30). It is possible to improve the javadoc but If we really want to make this _design choice_ explicit, I think we should create a new task for document it.
Btw it is not clear where we want to document this because we do not support "custom addon" in Kogito (yet?) so the way to load an addon class is an implementation detail and not an information that user should be aware of

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java (hunk 20 above)
Comment (kostola) [reply addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/event/EvaluateEvent.java:43]: The implementation of the addon was split in several tasks, as we discussed, because of its size and to have intermediate working "checkpoints" to validate our design.
This is the first one that implements a initial version of the addon that exports a limited amount of data.
I am already working on the second one that will refactor the exported data structure so that it won't expose any internal class of the DMN runtime anymore (which is your concern).
I agree that nobody should base any implementation on this `DMNResult`, but I don't think it's a big issue here because I don't think anyone will use this version of the addon since it will soon be enhanced.

File: addons/tracing/tracing-decision-common/src/test/resources/Traffic Violation.dmn (hunk 21 above)
Comment (kostola) [reply addons/tracing/tracing-decision-common/src/test/resources/Traffic Violation.dmn:1]: It's a test resource, so I think it's ok.

File: kogito-maven-plugin/src/main/java/org/kie/kogito/maven/plugin/GenerateModelMojo.java (hunk 24 above)
Comment (kostola) [reply kogito-maven-plugin/src/main/java/org/kie/kogito/maven/plugin/GenerateModelMojo.java:210]: Yes it's a good idea. We can open a dedicated ticket for this and extend the discussion to other Kogito devs.

File: pom.xml (hunk 25 above)
Comment (kostola) [reply pom.xml:758]: Yes now this sdk is being used in the addon.

File: addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java (hunk 19 above)
Comment (kostola) [reply addons/tracing/tracing-decision-quarkus-addon/src/main/java/org/kie/kogito/tracing/decision/KogitoDecisionTracingListener.java:25]: Let's keep this aligned with other classes in this PR so it can be merged.
I will take care of opening a dedicated ticket for enhancing the documentation.
